Implement fullscreen rendering update step

nt1m opened the pull request:
#### cba51d14a79420eea435792f37f3e379bc523bb9
<pre>
Implement fullscreen rendering update step
<a href="https://bugs.webkit.org/show_bug.cgi?id=249069">https://bugs.webkit.org/show_bug.cgi?id=249069</a>
<a href="https://rdar.apple.com/103209495">rdar://103209495</a>

Reviewed by Alex Christensen.

Event dispatching should be done by the rendering update steps.

<a href="https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering">https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering</a>
<a href="https://fullscreen.spec.whatwg.org/#run-the-fullscreen-steps">https://fullscreen.spec.whatwg.org/#run-the-fullscreen-steps</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-ready-allowed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-namespaces-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/move-to-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/move-to-inactive-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-root-block-scroll-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-root-block-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/ua-style-iframe-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-and-move-to-iframe-expected.txt: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-timing-expected.txt: Removed.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
(WebCore::FullscreenManager::resolvePendingPromise):
(WebCore::FullscreenManager::rejectPendingPromise):
(WebCore::FullscreenManager::dispatchPendingEvents):
(WebCore::FullscreenManager::notifyAboutFullscreenChangeOrError): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
* Source/WebCore/page/Quirks.cpp:
(WebCore::handleBBCQuirks):
(WebCore::Quirks::shouldDelayFullscreenEventWhenExitingPictureInPictureQuirk const): Deleted.
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: <a href="https://commits.webkit.org/289984@main">https://commits.webkit.org/289984@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d8a74aee590c903beb1ead43db553e26d44efc9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88606 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93566 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39360 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90657 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8512 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16309 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68316 "Found 2 new test failures: fullscreen/full-screen-enter-while-exiting.html media/media-fullscreen-not-in-document.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26026 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91608 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6512 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80135 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48680 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6278 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34548 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38468 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76647 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35443 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95406 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/15781 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11555 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16037 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75995 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76458 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18825 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20845 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19272 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/8820 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/15797 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15538 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18987 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17320 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -479,19 +480,22 @@ bool FullscreenManager::willEnterFullscreen(Element& element, HTMLMediaElementEn
 
     if (backForwardCacheState() != Document::NotInBackForwardCache) {
         ERROR_LOG(LOGIDENTIFIER, "Document in the BackForwardCache; bailing");
+        resolvePendingPromise();
```
Should this not be `rejectPendingPromise()`?

I see references to `m_pendingPromise->reject(Exception { ExceptionCode::TypeError });` removed; does this mean we never reject the fullscreen promise any longer?

nt1m replied:
> Should this not be rejectPendingPromise()?

Good catch! There might be a reason I decided to go with resolve (maybe tests?), but I guess I'll find out.

> I see references to m_pendingPromise->reject(Exception { ExceptionCode::TypeError }); removed; does this mean we never reject the fullscreen promise any longer?

I refactored those in https://github.com/WebKit/WebKit/commit/bb6320734c7fe75a57c2fd7abdc2673266ac19d1 , they're mostly happening from one method now

jernoble commented on Source/WebCore/dom/FullscreenManager.cpp:
```diff
@@ -760,14 +758,12 @@ void FullscreenManager::emptyEventQueue()
+    RefPtr target = document.fullscreenManager().fullscreenElement();
+    ASSERT(target);
+    document.fullscreenManager().addElementToChangeEventQueue(*target);
+    document.scheduleRenderingUpdate(RenderingUpdateStep::Fullscreen);
```
Lets not leave a future null pointer dereference in our code, even if there's an ASSERT.

jernoble commented on Source/WebCore/dom/FullscreenManager.h:
```diff
@@ -121,13 +121,16 @@ class FullscreenManager final : public CanMakeWeakPtr<FullscreenManager>, public
+
     Deque<GCReachableRef<Node>> m_fullscreenChangeEventTargetQueue;
     Deque<GCReachableRef<Node>> m_fullscreenErrorEventTargetQueue;
+    void addElementToChangeEventQueue(Node& target) { m_fullscreenChangeEventTargetQueue.append(GCReachableRef(target)); }
```
This should go up in the private functions, rather than mixed into the ivars.

nt1m committed:
Implement fullscreen rendering update step
https://bugs.webkit.org/show_bug.cgi?id=249069
rdar://103209495

Reviewed by Alex Christensen.

Event dispatching should be done by the rendering update steps.

https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering
https://fullscreen.spec.whatwg.org/#run-the-fullscreen-steps

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-ready-allowed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-namespaces-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/move-to-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/model/move-to-inactive-document-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-pseudo-class-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-root-block-scroll-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/fullscreen-root-block-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/fullscreen/rendering/ua-style-iframe-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-and-move-to-iframe-expected.txt: Added.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-timing-expected.txt: Removed.
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::FullscreenManager::requestFullscreenForElement):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::didExitFullscreen):
(WebCore::FullscreenManager::resolvePendingPromise):
(WebCore::FullscreenManager::rejectPendingPromise):
(WebCore::FullscreenManager::dispatchPendingEvents):
(WebCore::FullscreenManager::notifyAboutFullscreenChangeOrError): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateRendering):
* Source/WebCore/page/Quirks.cpp:
(WebCore::handleBBCQuirks):
(WebCore::Quirks::shouldDelayFullscreenEventWhenExitingPictureInPictureQuirk const): Deleted.
* Source/WebCore/page/Quirks.h:
* Source/WebCore/page/QuirksData.h:

Canonical link: https://commits.webkit.org/289984@main
